database/ydb/ydb.go

```diff
@@ -0,0 +1,311 @@
+		}
+	}
+	return options, nil
+}
+
+func (db *YDB) Close() error {
+	return db.driver.Close(context.TODO())
+}
+
+func (db *YDB) Run(migration io.Reader) error {
+	rawMigrations, err := io.ReadAll(migration)
+	if err != nil {
+		return err
+	}
+
+	res, err := db.driver.Scripting().Execute(context.TODO(), string(rawMigrations), nil)
+	if err != nil {
+		return err
+	}
+	return res.Close()
+}
+
+func (db *YDB) SetVersion(version int, dirty bool) error {
+	deleteVersionQuery := fmt.Sprintf(`
+		DELETE FROM %s 
+	`, db.config.MigrationsTable)
+
+	insertVersionQuery := fmt.Sprintf(`
+		INSERT INTO %s (version, dirty, created) VALUES (%d, %t, CurrentUtcTimestamp())
+	`, db.config.MigrationsTable, version, dirty)
+
+	ctx := context.TODO()
+	err := db.driver.Query().DoTx(ctx, func(ctx context.Context, tx query.TxActor) error {
+		if err := tx.Exec(ctx, deleteVersionQuery); err != nil {
+			return err
+		}
+		// Also re-write the schema version for nil dirty versions to prevent
+		// empty schema version for failed down migration on the first migration
+		// See: https://github.com/golang-migrate/migrate/issues/330
+		if version >= 0 || (version == database.NilVersion && dirty) {
+			if err := tx.Exec(ctx, insertVersionQuery); err != nil {
+				return err
+			}
+		}
+		return nil
+	}, query.WithTxSettings(query.TxSettings(query.WithSerializableReadWrite())))
+	return err
+}
+
+func (db *YDB) Version() (version int, dirty bool, err error) {
+	ctx := context.TODO()
+
+	getVersionQuery := fmt.Sprintf(`
+		SELECT version, dirty FROM %s ORDER BY version DESC LIMIT 1
+	`, db.config.MigrationsTable)
+
+	rs, err := db.driver.Query().QueryResultSet(ctx, getVersionQuery)
+	if err != nil {
+		return 0, false, &database.Error{OrigErr: err, Query: []byte(getVersionQuery)}
+	}
+	defer func() {
+		if closeErr := rs.Close(ctx); closeErr != nil {
+			err = multierror.Append(err, closeErr)
+		}
+	}()
+
+	row, err := rs.NextRow(ctx)
+	if err != nil {
+		if errors.Is(err, io.EOF) {
+			return database.NilVersion, false, nil
+		}
+		return 0, false, err
+	}
+
+	var v uint64
+	if err = row.Scan(&v, &dirty); err != nil {
+		return 0, false, &database.Error{OrigErr: err, Query: []byte(getVersionQuery)}
+	}
+	return int(v), dirty, err
+}
+
+func (db *YDB) Drop() (err error) {
+	ctx := context.TODO()
+
+	listQuery := "SELECT DISTINCT Path FROM `.sys/partition_stats` WHERE Path NOT LIKE '%/.sys%'"
+	rs, err := db.driver.Query().QueryResultSet(context.TODO(), listQuery)
+	if err != nil {
+		return &database.Error{OrigErr: err, Query: []byte(listQuery)}
+	}
+	defer func() {
+		if closeErr := rs.Close(ctx); closeErr != nil {
+			err = multierror.Append(err, closeErr)
+		}
+	}()
+
+	for {
+		var row query.Row
+		if row, err = rs.NextRow(ctx); err != nil {
+			if errors.Is(err, io.EOF) {
+				err = nil
+				break
+			}
+			return err
+		}
+
+		var table string
+		if err = row.Scan(&table); err != nil {
+			return err
+		}
+
+		dropQuery := fmt.Sprintf("DROP TABLE `%s`", table)
```

Comment:
Other db adapters in the repo use `DROP TABLE IF EXISTS`, I am not really sure why, but maybe we should keep it consistent and add IF EXISTS here